In [2]:
import os
import sys
sys.path.append(os.path.join(".."))

from evo_models import *

%load_ext autoreload
%autoreload 2

# Trening sieci MLP przy pomocy algorytmu ewolucyjnego

In [5]:
pop = ModelPopulation(
    # architecture=MLPArchitecture(7, [20, 10, 5], 1),
    architecture=MLPArchitecture(4, [5], 3),
    dataset_name="iris",
    last_layer_activation_func=Softmax(),
    loss_function="cross_entropy",
    population_size=100,
)
pop.show_test_loss()

Best model test loss: 0.8859929195943297


In [6]:
pop.train(epochs=100)

100%|█████████████| 100/100 [00:18<00:00] , Best Loss=0.4567


In [7]:
pop.train(epochs=100, mutation_strength=0.5)

100%|█████████████| 100/100 [00:18<00:00] , Best Loss=0.1081


In [8]:
pop.train(epochs=300, mutation_strength=0.5)

100%|█████████████| 300/300 [00:54<00:00] , Best Loss=0.0133


In [9]:
pop.train(epochs=500)

100%|█████████████| 500/500 [01:31<00:00] , Best Loss=0.0000


In [10]:
pop.train(epochs=500, mutation_strength=0.01)

100%|█████████████| 500/500 [01:31<00:00] , Best Loss=0.0000


In [11]:
pop.classification_performance_summary()

Model: model
Age: 0
Train Loss: 0.68
Test Loss: 0.0
Accuracy: 1
F1 Score: 1
Model made 75 / 75 correct predictions on the test set.
There were 0 incorrect predictions.


# Trening sieci MLP przy pomocy propagacji wstecznej

### Architektura (7, [20, 10, 5], 1)

In [11]:
model = MLP(
    architecture=MLPArchitecture(4, [20, 10, 5], 3),
    dataset_name="iris",
    last_layer_activation_func=Softmax(),
    loss_function="cross_entropy",
)

In [12]:
model.classification_performance_summary()

Model: model
Age: 0
Train Loss: 1.28
Test Loss: 1.45
Accuracy: 0.3067
F1 Score: 0.1439
Model made 23 / 75 correct predictions on the test set.
There were 52 incorrect predictions.


In [13]:
model.train(
    epochs=int(5000),
    learning_rate=0.001,
    optimizer="rmsprop",
    rms_beta=0.9,
    save_till_best=True,
)

Training from age 0: (best_loss_test: 0.07, loss_test: 0.14) : 100%|██████████| 5000/5000 [00:05<00:00, 975.65it/s] 

--------------------
Reverting to best model at age 1739
Loss on train set: 0.05	Loss on test set: 0.07
--------------------
Model training finished at age 1739 with loss 0.07 on test set


In [14]:
model.classification_performance_summary()

Model: model
Age: 1739
Train Loss: 0.05
Test Loss: 0.07
Accuracy: 0.9733
F1 Score: 0.9733
Model made 73 / 75 correct predictions on the test set.
There were 2 incorrect predictions.


### Architektura (7, [5], 1), czyli taka sama jak dla algorytmu ewolucyjnego

In [19]:
model2 = MLP(
    architecture=MLPArchitecture(4, [5], 3),
    dataset_name="iris",
    last_layer_activation_func=Softmax(),
    loss_function="cross_entropy",
)

In [20]:
model2.train(
    epochs=int(5000),
    learning_rate=0.001,
    optimizer="rmsprop",
    rms_beta=0.9,
    save_till_best=True,
)

Training from age 0: (best_loss_test: 0.06, loss_test: 0.06) : 100%|██████████| 5000/5000 [00:03<00:00, 1399.46it/s]

--------------------
Reverting to best model at age 4492
Loss on train set: 0.06	Loss on test set: 0.06
--------------------
Model training finished at age 4492 with loss 0.06 on test set


In [21]:
model2.classification_performance_summary()

Model: model
Age: 4492
Train Loss: 0.06
Test Loss: 0.06
Accuracy: 0.9733
F1 Score: 0.9733
Model made 73 / 75 correct predictions on the test set.
There were 2 incorrect predictions.


# Wnioski
W przypadku tego zbioru propagacja wsteczna działa bardzo dobrze. Nie była jednak w stanie przekroczyć wartości funkcji straty (cross entropy) poniżej 0.06. Oznacza to że ostateczny model osiąga wynik F1-score rzędu 0.97

Dla algorytmu ewolucyjnego szczególne korzyści przyniósł trening z większą siłą mutacji. Pozwolił on osiągnąć wyraźnie niższą wartość funkcji straty (~0.001)

Co więcej, model w dalszym ciągu wykazywał skłonność do nauki